### DISCUSSION TITLE
Accumulating Two-Stage Form Compiler memory

### DISCUSSION THREAD

**Discussion by lindsayad** [firedrakeproject/firedrake#3721] [Firedrake support]
Hi, we have an HDG Navier-Stokes application in which we perform a sequence of SNES solves with lower and lower viscosity using the previous solution as the initial guess for the next solve. I am running into OOM errors on my cluster. Using `tracemalloc`, the memory accumulation appears to be occurring due to
```
/home/lindad/firedrake/src/firedrake/firedrake/tsfc_interface.py:95: size=87.2 MiB, count=592773, average=154 B 
```
Do you have any suggestions on how to clear this memory? I added a `gc.collect()` at the end of each solve but that doesn't appear to have an effect on this.

**Comment by connorjward** [comment]
It might be the in-memory cache here: https://github.com/firedrakeproject/firedrake/blob/master/firedrake/tsfc_interface.py#L57. In which case you want to do

`TSFCKernel._cache.clear()`

this is what [Thetis do in their tests](https://github.com/thetisproject/thetis/blob/master/test/conftest.py).

**Reply by lindsayad** [reply]
Unfortunately that did not fix it. I wish I could see more of a stack trace from `tracemalloc` than
```
196380 memory blocks: 29662.4 KiB
  File "/home/lindad/firedrake/src/firedrake/firedrake/tsfc_interface.py", line 95
    return cls._cache.setdefault((key, comm.py2f()), pickle.loads(val))
```

**Comment by wence-** [comment]
These parameters shouldn't change the form (if you wrapped them in Constant), so you wouldn't make more kernels.

**Reply by lindsayad** [reply]
> These parameters

Are you referring to the viscosity?

**Reply by wence-** [reply]
For example, yes (assuming it's a scalar)

**Reply by lindsayad** [reply]
Applying
```
diff --git a/codes-hdg/navier-stokes/ns-hdgw.py b/codes-hdg/navier-stokes/ns-hdgw.py
index 9e70973..e699cdb 100644
--- a/codes-hdg/navier-stokes/ns-hdgw.py
+++ b/codes-hdg/navier-stokes/ns-hdgw.py
@@ -144,12 +144,12 @@ def run():
             pex = sin(pi*(x-1))*cos(pi*(y-1))
             pex = pex - assemble(pex*dx)
             uex = as_vector([ux, uy])
-            nu = char_U * char_L / re
+            nu = Constant(char_U * char_L / re)
             f = -2*nu*div(sym(grad(uex))) + grad(pex) + dot(grad(uex), uex)
             dirlist = [1, 2, 3, 4]
             solex = [uex, pex]
 
-        nu = char_U * char_L / re
+        nu = Constant(char_U * char_L / re)
         if rank == 0:
             print("\n\n *** Navier-Stokes with Reynolds = %5d and nu = %3.3e *** \n\n" % (re, nu), flush=True)
 
```
leads to the error
```
Traceback (most recent call last):
  File "/home/lindad/firedrake/src/PyOP2/pyop2/global_kernel.py", line 349, in __call__
    func = self._func_cache[key]
           ~~~~~~~~~~~~~~~~^^^^^
KeyError: 140397825045904
```

**Reply by lindsayad** [reply]
Do you guys have any ideas regarding this failure? I've seen this error in the past and I've been able to avoid it by tweaking my Reynolds number but now I can't seem to avoid the error

**Reply by wence-** [reply]
Can you show the whole backtrace? Also if you have a self-contained example that reproduces it, it will be easier to see what is going on.

**Reply by lindsayad** [reply]
I am pretty certain the cache error cropped up due to some corruption of my PETSC_DIR/PETSC_ARCH environment. I found that I had errors like these during compilation of cache objects:
```
In file included from /home/lindad/petsc/include/petscbag.h:4,                                                            
                 from /home/lindad/petsc/include/petsc.h:5,                                                               
                 from /home/lindad/firedrake/.cache/pyop2/7a/af4e38fefb6cc499b6d8cf9076eb2a_p1259075.c:3:                 
/home/lindad/petsc/include/petscsys.h:67:24: error: unknown type name ‘PETSC_C_INLINE’                                    
   67 | #  define PETSC_INLINE PETSC_C_INLINE                                                                             
      |                        ^~~~~~~~~~~~~~                                                                             
/home/lindad/petsc/include/petscsys.h:70:36: note: in expansion of macro ‘PETSC_INLINE’                                   
   70 | #define PETSC_STATIC_INLINE static PETSC_INLINE                                                                   
      |                                    ^~~~~~~~~~~~                                                                   
/home/lindad/petsc/include/petscmath.h:396:1: note: in expansion of macro ‘PETSC_STATIC_INLINE’                           
  396 | PETSC_STATIC_INLINE PetscReal PetscAbsScalar(PetscScalar a) {return a < 0.0 ? -a : a;}                            
      | ^~~~~~~~~~~~~~~~~~~                                                                                               
In file included from /home/lindad/petsc/include/petscsys.h:328:
```
After rebuilding my environment I am not getting these errors anymore

**Reply by lindsayad** [reply]
No longer getting the `_func_cache` errors, I'm still seeing the growth from the `tsfc_interface` every solve even with wrapping $nu$ in `Constant`. So I must still be doing something wrong

**Reply by lindsayad** [reply]
What do you guys use for memory profiling? I know your site suggests using `memory_profiler`. However, I am not finding its output (attached) to be helpful in tracking down where all the memory allocation is. And `tracemalloc` just gives me a single frame for the `tsfc_interface.py`.

[profiler.log](https://github.com/user-attachments/files/16503337/profiler.log)

**Reply by connorjward** [reply]
I don't do a tremendous amount of memory profiling so I don't know another tool to recommend. @Ig-dolci you have been debugging some memory leaks recently, do you have any memory-profiling tool recommendations?

Looking through your code it looks like you reuse the same mesh hierarchy for each problem. Ideally that should be fine but we do cache a fair amount of things on the mesh so this might be where you are running out of room. It may be informative to try moving the mesh construction code into the loop to see if that stops this issue, which would point us to where the problem is.

**Reply by Ig-dolci** [reply]
I am using `memory_profiler`, which has worked for my memory leak investigation.  @lindsayad Are you using `NonlinearVariationalSolver`?

**Reply by lindsayad** [reply]
Hi @Ig-dolci, yes I am
